src/expr/symbol_table.cpp

```diff
@@ -255,18 +314,121 @@ void SymbolTable::Implementation::reset() {
   new (this) SymbolTable::Implementation();
 }
 
+void SymbolTable::Implementation::bindWithOverloading(const string& name, Expr obj) {
+  CDHashMap<string, Expr>::const_iterator it = d_exprMap->find(name);
+  if(it != d_exprMap->end()) {
+    if((*it).second!=obj) {
+      if(!isOverloadedFunction((*it).second)) {
+        // mark previous as overloaded
+        markOverloaded(name, (*it).second);
+      }
+      // mark this as overloaded
+      markOverloaded(name, obj);
+    }
+  }
+}
+
+void SymbolTable::Implementation::markOverloaded(const string& name, Expr obj) {
+  Trace("parser-overloading") << "Overloaded function : " << name;
+  Trace("parser-overloading") << " with type " << obj.getType() << std::endl;
+  d_overloaded_symbols->insert(obj);
+  // get the argument types
+  Type t = obj.getType();
+  Type rangeType = t;
+  std::vector< Type > argTypes;
+  if(t.isFunction()) {
+    argTypes = ((FunctionType)t).getArgTypes();
```

Comment:
Here and below: use static_cast<type>(expr) instead of the C casts ((type)expr).  This is covered in the style guide (though admittedly in a later section that we haven't discussed yet).